spinalcordtoolbox/deepseg_gm/deepseg_gm.py

```diff
@@ -257,29 +206,24 @@ def segment_volume(ninput_volume, model_name,
         for i in range(8):
             sampled_value = np.random.uniform(high=2.0)
             sampled_axial_slices = axial_slices + sampled_value
-            preds = deepgmseg_model.predict(sampled_axial_slices,
-                                            batch_size=BATCH_SIZE,
-                                            verbose=True)
+            preds = ort_sess.run(output_names=["predictions"], input_feed={"input_1": sampled_axial_slices})[0]
```

Comment:
Aha! It looks like adding that `batch_processing.sh` test for Windows (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3770) is immediately coming in handy:

The `.ONNX` model now throws the following error (as per [this CI run](https://github.com/spinalcordtoolbox/spinalcordtoolbox/runs/6474474974?check_suite_focus=true)):

```
# Segment gray matter
sct_deepseg_gm -i t2s.nii.gz -qc "$SCT_BP_QC_FOLDER"

--
Spinal Cord Toolbox (git-jn/3735-replace-tensorflow-with-onnxruntime-65e127fa754fc4da5a6d9ba4aebf1ba340ca2910)

sct_deepseg_gm -i t2s.nii.gz -qc D:/a/spinalcordtoolbox/spinalcordtoolbox/spinalcordtoolbox/qc_example_data
--

2022-05-17 16:19:48.7539231 [E:onnxruntime:, sequential_executor.cc:339 onnxruntime::SequentialExecutor::Execute] Non-zero status code returned while running Concat node. Name:'concatenate_1' Status Message: bad allocation
Traceback (most recent call last):
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\scripts\sct_deepseg_gm.py", line 136, in <module>
    main(sys.argv[1:])
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\scripts\sct_deepseg_gm.py", line 119, in main
    use_tta)
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\deepseg_gm\deepseg_gm.py", line 257, in segment_file
    use_tta)
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\deepseg_gm\deepseg_gm.py", line 218, in segment_volume
    preds = ort_sess.run(output_names=["predictions"], input_feed={"input_1": axial_slices})[0]
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\venv_sct\lib\site-packages\onnxruntime\capi\onnxruntime_inference_collection.py", line 188, in run
    return self._sess.run(output_names, input_feed, run_options)
onnxruntime.capi.onnxruntime_pybind11_state.RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running Concat node. Name:'concatenate_1' Status Message: bad allocation
```

Reply:
I'm going to make the PR as draft until I can resolve this issue.

Reply:
I've spent the entire day debugging, and I have some interesting takeaways:

### 1. The memory issues are intermittent

I restarted the job from this PR branch for a third time, and it passed. (So, the memory issues are at least _somewhat_ intermittent -- though, from my testing, I'd say 95% of jobs failed as-is.)

### 2. The memory issues depend on how `sct_deepseg_gm` is called

If I call `sct_deepseg_gm` using `sct_deepseg_gm.main()`, the script runs without error.

```python
    sct_deepseg_gm.main(["-i", os.path.join("data", "sct_example_data",
                                            "t2s", "t2s.nii.gz")])
```

However, if I call `sct_deepseg_gm` using `subprocess.run`:

```python
# Call the python scripts directly through subprocess (to mimic compat/launcher.py)
subprocess.run([sys.executable,
                sct_dir_local_path("spinalcordtoolbox", "scripts", "sct_deepseg_gm.py"),
                "-i", os.path.join("data", "sct_example_data",
                                   "t2s", "t2s.nii.gz")])

# Call the SCT CLI command itself (i.e. using compat/launcher.py)
subprocess.run(["sct_deepseg_gm",
                "-i", os.path.join("data", "sct_example_data",
                                   "t2s", "t2s.nii.gz")])
```

Then _numpy_ will throw the error instead:

```
numpy.core._exceptions.MemoryError: Unable to allocate 12.5 MiB for an array with shape (32, 640, 640) and data type uint8
```

### 3. The `numpy` error leads us to better memory-related advice

Trying to search on `bad allocation onnxruntime` wasn't bringing up much of use. But, searching on the numpy error yielded _much_ more fruitful results: https://stackoverflow.com/q/57507832

For Linux:

> This is likely due to your system's [overcommit handling](https://www.kernel.org/doc/Documentation/vm/overcommit-accounting) mode.
>
> In the default mode, 0,
>
>    Heuristic overcommit handling. Obvious overcommits of address space are refused. Used for a typical system. It ensures a seriously wild allocation fails while allowing overcommit to reduce swap usage. The root is allowed to allocate slightly more memory in this mode. This is the default.

For Windows:

> I had this same problem on Window's and came across this solution. So if someone comes across this problem in Windows the solution for me was to increase the [pagefile](https://whatis.techtarget.com/definition/pagefile) size, as it was a Memory overcommitment problem for me too.

### 4. Increasing the pagefile size seems to fix the issue

Following the advice above, I used the [`configure-pagefile-action`](https://github.com/al-cheb/configure-pagefile-action) to set a larger pagefile:

```yaml
    steps:
    - name: configure Pagefile
      uses: al-cheb/configure-pagefile-action@v1.2
      with:
        minimum-size: 8GB
```

This fixed the issue. ([Before](https://github.com/spinalcordtoolbox/spinalcordtoolbox/runs/6583354152?check_suite_focus=true#step:14:45), [After](https://github.com/spinalcordtoolbox/spinalcordtoolbox/runs/6583358765?check_suite_focus=true#step:15:33))

----

I'm still a little fuzzy on the concept of memory overcommitment on Windows, so I'm not sure if this represents an acceptable solution or not. I would have to try and study a little more to make sure I'm not just sweeping the root cause under the rug here.

Reply:
This issue may be of use when it comes to profiling the memory usage of `onnxruntime`: https://github.com/microsoft/onnxruntime/issues/7067

Reply:
Using `memory_profiler`, we can see that when calling `ort_sess.run`, memory usage balloons  from ~200MB to ~6GB:

```python
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   172    206.0 MiB    206.0 MiB           1   @profile
   173                                         def onnx_prediction(model_abs_path, input):
   174    206.0 MiB      0.0 MiB           1       sess_options = ort.SessionOptions()
   175    206.1 MiB      0.0 MiB           1       sess_options.enable_profiling = True
   176    212.6 MiB      6.6 MiB           1       ort_sess = ort.InferenceSession(model_abs_path, sess_opti
ons=sess_options)
   177   5792.0 MiB   5579.3 MiB           1       preds = ort_sess.run(output_names=["predictions"], input_
feed={"input_1": input})[0]
   178   5792.0 MiB      0.0 MiB           1       return preds
```

Searching for `onnxruntime` options to reduce memory usage, I found `enable_cpu_mem_arena`. Setting this to false completely addresses the issue:

```python
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   172    206.0 MiB    206.0 MiB           1   @profile
   173                                         def onnx_prediction(model_abs_path, input):
   174    206.1 MiB      0.0 MiB           1       sess_options = ort.SessionOptions()
   175    206.1 MiB      0.0 MiB           1       sess_options.enable_profiling = True
   176    206.1 MiB      0.0 MiB           1       sess_options.enable_cpu_mem_arena = False
   177    212.4 MiB      6.4 MiB           1       ort_sess = ort.InferenceSession(model_abs_path, sess_opti
ons=sess_options)
   178    217.8 MiB      5.3 MiB           1       preds = ort_sess.run(output_names=["predictions"], input_
feed={"input_1": input})[0]
   179    217.8 MiB      0.0 MiB           1       return preds
```

From the docs on [`enable_cpu_mem_area`](https://onnxruntime.ai/docs/api/python/api_summary.html#onnxruntime.SessionOptions.enable_cpu_mem_arena):

> Enables the memory arena on CPU. Arena may pre-allocate memory for future usage. Set this option to false if you don’t want it. Default is True.

I still don't exactly understand the inner-workings here (I'd like to ask the `onnxruntime` devs a little more about what's happening here), but at least we're able to pin down the issue, and a potential solution.

Reply:
Implemented in [196cc04](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3738/commits/196cc04d14cf5b59decb542635dcf1dc6a2341f1).